### PR TITLE
fix recent regression: macro paradise + Java sources = MatchError

### PR DESCRIPTION
**Pull request by SethTisue**
fixes https://github.com/scala/scala-dev/issues/303 ;
the regression was in https://github.com/scala/scala/pull/5585

### CODE REVIEW

**Comment by SethTisue** [comment]
this is still WIP until I do a community build run to verify that it fixes the problem

testing locally with `; setupPublishCore ; publishM2` from the scala repo followed by `version=2.12.2-c8e5504-SNAPSHOT ./run.sh breeze,monix,akka-http` on the 2.12.x branch of scala/community-builds

and https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1269/consoleFull

**Comment by SethTisue** [comment]
the MiMa and ScalaCheck failures are also seen at https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-bootstrap/806/consoleFull and don't have to do with this PR

the MiMa thing should go away if I rebase after https://github.com/scala/scala/pull/5712 is merged

the ScalaCheck thing is https://github.com/scala/scala-dev/issues/296

**Comment by SethTisue** [comment]
oops, previous community build run used the wrong SHA. try again: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/1270/consoleFull (404 til Jenkins gets caught up)

**Comment by SethTisue** [comment]
local community build run passed, so this is ready for review

File: src/reflect/scala/reflect/internal/ReificationSupport.scala [hunk 1]
```diff
@@ -351,9 +351,13 @@ trait ReificationSupport { self: SymbolTable =>
                                        List[Tree], List[Tree], ValDef, List[Tree])] = tree match {
         case ClassDef(mods, name, tparams, impl) =>
           val X = if (mods.isCase) UnMkTemplate.asCase else UnMkTemplate
-          val X(parents, selfType, ctorMods, vparamss, earlyDefs, body) = impl
-          if (ctorMods.isTrait || ctorMods.hasFlag(JAVA)) None
-          else Some((mods, name, tparams, ctorMods, vparamss, earlyDefs, parents, selfType, body))
+          impl match {
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/ReificationSupport.scala:354]: How about negating the condition, moving it into the case guard and collapsing then else None branch into the default case?

**Review by adriaanm — CHANGES_REQUESTED** [review]
Yay for a green community build! Small suggestion for branch thrift.

File: src/reflect/scala/reflect/internal/ReificationSupport.scala (hunk 1 above)
Comment (SethTisue) [reply src/reflect/scala/reflect/internal/ReificationSupport.scala:354]: Okay, but I charge extra whenever I have to use De Morgan's law.

Comment (adriaanm) [reply src/reflect/scala/reflect/internal/ReificationSupport.scala:354]: IntelliJ has it built-in :-p

**Review by adriaanm — APPROVED** [review]

File: src/reflect/scala/reflect/internal/ReificationSupport.scala (hunk 1 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/ReificationSupport.scala:354]: Definitely use Frank Morgan's law here. The one about not looking behind the curtain. BTW, his real name was Wuppermann. I definitely want the t-shirt with the big W on it. Sorry I didn't arrive in time for this review.

**Pull request merged**